[kernel 2a/n] Split hashing/index `GetUTXOStats` codepaths, decouple from `coinstatsindex`

dongcarl opened the pull request:
Part of: #24303
Depends on: #24322

The `GetUTXOStats` function has 2 codepaths:
  - One which queries the `CoinStatsIndex` for the UTXO hash
  - One which actually performs the hashing

For `libbitcoinkernel`, the only place where we call `GetUTXOStats` is in `PopulateAndValidateSnapshots`, which uses the `SHA256D` hash, and is therefore unable to use the `CoinStatsIndex` since that only provides `MuHash` hashes. Not that I think indices necessarily belong in `libbitcoinkernel` anyway.

This PR separates these 2 aforementioned codepaths of `GetUTXOStats`, uses the hashing codepath in `PopulateAndValidateSnapshots`, and removes the need to link in `index/coinstatsindex.cpp` and `node/coinstats.cpp`.

-----

Logistically, this PR:
- Extracts out the `index_requested` and `hash_type` members of `CoinStats`, which served as "in-params" to `GetUTXOStats` embedded within the `CoinStats` struct. This allows `CoinStats` to only consist of "out-param" members, and be returned by `GetUTXOStats` without needing to be an "in-out" param
- Introduce the purely virtual `UTXOHashers` class, with 3 implementations: `SHA256DHasher`, `MuHashHasher`, and `NullHasher`. These replace the existing template-based polymorphism.
- Split `GetUTXOStats` into:
    - `CalculateUTXOStatsWithHasher(UTXOHasher&, ...)`, and
    - `LookupUTXOStatsWithIndex(CoinStatsIndex&, ...)`
- Use `CalculateUTXOStatsWithHasher` directly where appropriate (`src/validation.cpp` and `src/fuzz`)
- Move `GetUTXOStats` to `rpc/blockchain`, which is the only place that depends on `GetUTXOStats`'s weird fallback behaviour
- Move `LookupUTXOStatsWithIndex` to `index/coinstatsindex`

Code organization:
- `src/`
  - `kernel/` → only contains the hashing codepath
    - `coinstats.cpp` → hashing codepath implementations
    - `coinstats.h` → header for `kernel/coinstats.cpp`
  - `index/` → only contains the index codepath
    - `coinstatsindex.cpp` → index codepath implementations
    - `coinstatsindex.h`
  - `validation.cpp` → only uses the hashing codepath
  - `rpc/blockchain.cpp` → uses both the hashing and index codepath, old `GetUTXOStats` fallback logic moved here as static
  - `test/fuzz/coins_view.cpp` → only uses the hashing codepath


TODOs:
- [x] Commit messages could be fleshed out more

Would love any feedback!

dongcarl committed:
coinstats: Extract hash_type in-member to in-param

Currently, CCoinsStats is a struct with both in-params and out-params
where the hash_type and index_requested members are the only in-params.

This change removes CCoinsStats' hash_type in-param member and adds it
to the relevant functions instead.

[META] In subsequent commits, all of CCoinsStats' members which serve as
       in-params will be moved out so as to make CCoinsStats a pure
       out-param struct.

dongcarl committed:
coinstats: Extract index_requested in-member to in-param

This change removes CCoinsStats' index_requested in-param member and
adds it to the relevant functions instead.

dongcarl committed:
coinstats: Return purely out-param CCoinsStats

In previous commits in this patchset, we removed all in-param members of
CCoinsStats. Now that that's done, we can modify GetUTXOStats to return
an optional CCoinsStats instead of a status bool. Callers are modified
accordingly.

In rpc/blockchain.cpp, we discover that GetUTXOStats' status bool when
getting UTXO stats for pprev was not checked for error. We fix this as
well.

dongcarl committed:
coinstats: Separate hasher/index lookup codepaths

Split out ComputeUTXOStats and LookupUTXOStatsWithIndex from
GetUTXOStats, since the hashing and indexing codepaths are quite
disparate in practice.

Also allow add a constructor to CCoinsStats for it to be constructed
from a a block height and hash. This is used in both codepaths.

Also add a note in GetUTXOStats documenting a behaviour quirk that
predates this patchset.

[META] This allows the hashing codepath to be moved to a separate file
       in a future commit, decoupling callers that only rely on the
       hashing codepath from the indexing one. This is key for
       libbitcoinkernel, which needs to have the hashing codepath for
       AssumeUTXO, but does not wish to be coupled with indexes.

dongcarl committed:
kernel: Use ComputeUTXOStats in validation

This is the "fruit of our labor" for this patchset.
ChainstateManager::PopulateAndValidateSnapshot can now directly call
ComputeUTXOStats(...).

Our consensus engine is now fully decoupled from all indices.

See the src/Makefile.am for some satisfying removals.

dongcarl committed:
coinstats: Move hasher codepath to kernel/coinstats

As mentioned in a previous commit, the hashing codepath can now be moved
to a separate file. This decouples callers that only rely on the hashing
codepath from the indexing one.

This is key for libbitcoinkernel, which needs to have the CoinsStats
hashing codepath for AssumeUTXO, but does not wish to be coupled with
indexes.

Note that only the .cpp file is split in this commit, the header files
will be split in a subsequent commit and the #includes to
node/coinstats.h will be adjusted to only #include the necessary
headers.

dongcarl committed:
coinstats: Split node/coinstats.h to kernel/coinstats.h

Most of this commit is pure-move.

After this change:

- kernel/coinstats.h
    -> Contains declarations for:
       - enum class CoinStatsHashType
       - struct CCoinsStats
       - GetBogoSize(...)
       - TxOutSer(...)
       - ComputeUTXOStats(...)
- node/coinstats.h
    -> Just GetUTXOStats, which will be removed as we change callers to
       directly use the hashing/indexing codepaths in future commits.

dongcarl committed:
scripted-diff: Move src/kernel/coinstats to kernel::

Introduces a new kernel:: namespace and move all of src/kernel/coinstats
under it.

In the verify script, lines like:

line="$(grep -n 'namespace node {' -- src/kernel/coinstats.h | tail -n1 | cut -d: -f1)"
sed -i -e "${line}s@namespace node {@namespace kernel {@" -- src/kernel/coinstats.h

Are intended to replace only the last instance of "namespace node" with
"namespace kernel", this is to avoid replacing forward declarations of
things inside the node:: namespace.

-BEGIN VERIFY SCRIPT-
sed -E -i 's@namespace node@namespace kernel@g' -- src/kernel/coinstats.cpp

line="$(grep -n 'namespace node {' -- src/kernel/coinstats.h | tail -n1 | cut -d: -f1)"
sed -i -e "${line}s@namespace node {@namespace kernel {@" -- src/kernel/coinstats.h

line="$(grep -n '// namespace node' -- src/kernel/coinstats.h | tail -n1 | cut -d: -f1)"
sed -i -e "${line}s@// namespace node@// namespace kernel@" -- src/kernel/coinstats.h

things='(CCoinsStats|CoinStatsHashType|GetBogoSize|TxOutSer|ComputeUTXOStats)'
git grep -lE 'node::'"$things" | xargs sed -E -i 's@node::'"$things"'@kernel::\1@g'
sed -E -i 's@'"$things"'@kernel::\1@g' -- src/node/coinstats.cpp src/node/coinstats.h
sed -E -i 's@BlockManager@node::\0@g' -- src/kernel/coinstats.cpp
-END VERIFY SCRIPT-

dongcarl committed:
coinstats: Move GetUTXOStats to rpc/blockchain

rpc/blockchain.cpp is now the only user of the vestigial
GetUTXOStats(...). And since GetUTXOStats(...)'s special fallback logic
was only really relevant/meant for rpc/blockchain.cpp, we can just move
it there.

dongcarl committed:
style-only: Rearrange using decls after scripted-diff

ryanofsky reviewed (commented):
Approach ACK (a7883ac7a497f3d1cb52f5f71e71de5e5e9301f1). Looked through various commits and I think the changes all make sense

MarcoFalke commented:
Would an alternative be to treat assumeutxo as not part of the kernel for now (and completely move it out?)

dongcarl commented:
@MarcoFalke 

> Would an alternative be to treat assumeutxo as not part of the kernel for now (and completely move it out?)

Sure, but I don't prefer that alternative. I think it would actually be more work to extract assumeutxo from the kernel, plus if you look at the `GetUTXOStats` function and friends prior to this PR, it needed quite some cleaning up, which is done in this PR.

jonatack commented:
Concept ACK

willcl-ark commented:
Perhaps it's just my system, but the `feature_coinstatsindex.py` bitcoind does not get terminated after a successful run. This causes my script running tests against each commit to fail. All the other tests terminate just fine, and this happens running all tests and the test individually.

The logs appear to show the test and node sutting down, but this does not appear to be the case in practice.

```log
[I] [11:19] ubuntu:bitcoin (2022-02-libbitcoinkernel-p2-coinstats-minimal) | ₿ test/functional/feature_coinstatsindex.py
2022-04-29T10:20:14.316000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_so5w4sxf
2022-04-29T10:20:15.704000Z TestFramework (INFO): Test that gettxoutsetinfo() output is consistent with or without coinstatsindex option
2022-04-29T10:20:15.758000Z TestFramework (INFO): Test that gettxoutsetinfo() can get fetch data on specific heights with index
2022-04-29T10:20:15.850000Z TestFramework (INFO): Test gettxoutsetinfo() with index and verbose flag
2022-04-29T10:20:15.963000Z TestFramework (INFO): Test that the index is robust across restarts
2022-04-29T10:20:16.419000Z TestFramework (INFO): Test that the index works with -reindex
2022-04-29T10:20:17.198000Z TestFramework (INFO): Test that -reindex-chainstate is disallowed with coinstatsindex
2022-04-29T10:20:17.363000Z TestFramework (INFO): Test use_index option for nodes running the index
2022-04-29T10:20:17.450000Z TestFramework (INFO): Test that index can handle reorgs
> /home/will/src/bitcoin/test/functional/feature_coinstatsindex.py(55)run_test()
-> self._test_index_rejects_hash_serialized()
(Pdb) l
 50             self.wallet = MiniWallet(self.nodes[0])
 51             self._test_coin_stats_index()
 52             self._test_use_index_option()
 53             self._test_reorg_index()
 54             import pdb; pdb.set_trace()
 55  ->         self._test_index_rejects_hash_serialized()
 56
 57         def block_sanity_check(self, block_info):
 58             block_subsidy = 50
 59             assert_equal(
 60                 block_info['prevout_spent'] + block_subsidy,
(Pdb) n
2022-04-29T10:20:38.130000Z TestFramework (INFO): Test that the rpc raises if the legacy hash is passed with the index
--Return--
> /home/will/src/bitcoin/test/functional/feature_coinstatsindex.py(55)run_test()->None
-> self._test_index_rejects_hash_serialized()
(Pdb) n
> /home/will/src/bitcoin/test/functional/test_framework/test_framework.py(156)main()
-> exit_code = self.shutdown()
(Pdb) n
2022-04-29T10:20:43.555000Z TestFramework (INFO): Stopping nodes
2022-04-29T10:20:43.657000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_so5w4sxf on exit
2022-04-29T10:20:43.657000Z TestFramework (INFO): Tests successful
> /home/will/src/bitcoin/test/functional/test_framework/test_framework.py(157)main()
-> sys.exit(exit_code)
(Pdb) n
SystemExit: 0
> /home/will/src/bitcoin/test/functional/test_framework/test_framework.py(157)main()
-> sys.exit(exit_code)
(Pdb) c
[I] [11:20] ubuntu:bitcoin (2022-02-libbitcoinkernel-p2-coinstats-minimal) | ₿ ps -Af | rg bitcoind
will        6695    3480  0 Apr28 ?        00:14:46 bitcoind
will      586063    3480  0 11:20 pts/8    00:00:00 /home/will/src/bitcoin/src/bitcoind -datadir=/tmp/bitcoin_func_test_so5w4sxf/node1 -logtimemicros -debug -debugexclude=libevent -debugexclude=leveldb -uacomment=testnode1 -logthreadnames -logsourcelocations -coinstatsindex -reindex
```

dongcarl committed:
fuzz: Remove useless GetUTXOStats fuzz case

In the GetUTXOStats fuzz case, GetUTXOStats is always called with a
CCoinsViewCache. Which is guaranteed to throw a std::logic_error when
its ::Cursor() method is called on the first line of GetUTXOStats.

In the fuzz case, we basically catch this logic error and declare
victory if we caught it.

There is no point to fuzzing this deterministic logic.

Confirmed with IWYU that the node/coinstats.h #include is no longer
necessary.

dongcarl commented:
@willcl-ark Interesting observation but I tested it on master and I see the same behaviour there too! So this is unrelated to this PR's changes. Perhaps you'd like to open a separate Issue about this?

mzumsande commented:
> @willcl-ark Interesting observation but I tested it on master and I see the same behaviour there too! (you should verify!) So this is unrelated to this PR's changes. Perhaps you'd like to open a separate Issue about this?

That was an oversight by me in #24789. I opened #25034 to fix this.

willcl-ark commented:
> > @willcl-ark Interesting observation but I tested it on master and I see the same behaviour there too! (you should verify!) So this is unrelated to this PR's changes. Perhaps you'd like to open a separate Issue about this?
> 
> That was an oversight by me in #24789. I opened #25034 to fix this.

Ah thanks Martin, that would explain it. I will cherry-pick that and re-run then :)

willcl-ark commented:
@mzumsande cherry-picking https://github.com/bitcoin/bitcoin/commit/5345f752313c6bd7164109bdfd64ca3a55647c22 does indeed fix and allow all tests to pass at each commit in this PR

dongcarl commented:
@willcl-ark Okay for me to hide your comments so as to not confuse future reviewers?

willcl-ark commented:
> @willcl-ark Okay for me to hide your comments so as to not confuse future reviewers?

Please go ahead!

dongcarl commented:
Pushed 756f1a755d138e36a9e50e7e4e46ae0bcd3975ca -> 30503c98a8bc97b582f072c34168b5279d65a492
- Added commit "fuzz: Remove useless GetUTXOStats fuzz case", adjusted subsequent commits
- Removed some extraneous whitespace in scripted-diff

-----

Reviewers: Please let me know if the new "fuzz: Remove useless GetUTXOStats fuzz case" commit looks okay.

Ping @practicalswift

mzumsande reviewed (commented):
Concept ACK

I was able to follow the commits step by step and the changes look sensible to me. One suggestion below about the coinstatsindex codepath.

dongcarl committed:
Move logic from LookupUTXOStatsWithIndex to CoinStatsIndex::LookUpStats

The indexing codepath logic in node/coinstats.cpp is simple enough to be
moved into CoinStatsIndex::LookUpStats, avoiding an additional layer of
function calls. Callers are modified accordingly.

Also, add 2 missed BOOST_CHECKs to the coinstatsindex_initial_sync unit
test.

dongcarl committed:
Use only kernel/coinstats.h in index/coinstatsindex.h

Removes a circular dependency, horray!

dongcarl committed:
includes: Remove rpc/util.h -> node/coinstats.h

Confirmed with IWYU that this is unnecessary.

dongcarl commented on src/index/coinstatsindex.h:
```diff
@@ -62,4 +66,7 @@ class CoinStatsIndex final : public BaseIndex
 /// The global UTXO set hash object.
 extern std::unique_ptr<CoinStatsIndex> g_coin_stats_index;
 
+std::optional<kernel::CCoinsStats> GetUTXOStatsWithIndex(CoinStatsIndex& coin_stats_index, const CBlockIndex* pindex);
```
Great idea! See: 7dbebb3a9c4ea691b11659f2b91e1e4b8e847251

dongcarl commented on src/rpc/blockchain.cpp:
```diff
@@ -922,9 +924,13 @@ static RPCHelpMan gettxoutsetinfo()
+                if (maybe_prev_stats.has_value()) {
+                    prev_stats = maybe_prev_stats.value();
+                } else {
+                    throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to read UTXO set");
```
Right, I added some info in the commit message. Will be making the commit messages better in the coming days! 2155489fe6d103bdca26c0875058d563a3e8d5c0

dongcarl replied:
Commit messages are now complete! 😄

dongcarl commented:
Pushed 30503c98a8bc97b582f072c34168b5279d65a492 -> cfbe8efb08e42d404306c5d59d4a2e0719b2a788
- Merged `GetUTXOStatsWithIndex()` logic into `CoinStatsIndex::LookUpStats()` as mentioned here: https://github.com/bitcoin/bitcoin/pull/24410#discussion_r862389785 (Thanks Martin!)

dongcarl committed:
kernel: Remove unnecessary blockfilter{index,}.cpp

It is no longer necessary to link in blockfilter.cpp and
index/blockfilterindex.cpp after merge of PR#21726 since validation has
been decouple from the blockfilterindex.

dongcarl commented:
Pushed cfbe8efb08e42d404306c5d59d4a2e0719b2a788 -> 06dea7ea36eb1606cd4255e2875d0e4fe2fdd5a4
- Rebased over master for merge of #25034
- Fleshed out commit messages
- Added previously missing (even before this PR) `BOOST_CHECK`s in "Move logic from GetUTXOStatsWithIndex to CoinStatsIndex::LookUpStats" (b3d18d7bfcfeba7ef3ead5de0ee9bd58300609e9)
- Moved more declarations out of the "public header" in "coinstats: Split node/coinstats.h to kernel/**.h" (c4ab9ff40f1499f4abd2dd1915733c86bc4bccbf)
- Added commit at the beginning: "kernel: Remove unnecessary index/blockfilterindex.cpp" to account for merge of #21726

Miminlaili commented on src/index/coinstatsindex.cpp:
```diff
@@ -12,10 +12,11 @@
-using node::TxOutSer;
 using node::UndoReadFromDisk;
 
 static constexpr uint8_t DB_BLOCK_HASH{'s'};
```
[]()

Miminlaili commented on src/index/coinstatsindex.cpp:
```diff
@@ -316,28 +317,31 @@ static bool LookUpOne(const CDBWrapper& db, const CBlockIndex* block_index, DBVa
     return db.Read(DBHashKey(block_index->GetBlockHash()), result);
```
```suggestion
     return db.Read(DBHashKey(block_index->GetBlockHash()), result);
```

Miminlaili commented on src/index/coinstatsindex.cpp:
```diff
@@ -316,28 +317,31 @@ static bool LookUpOne(const CDBWrapper& db, const CBlockIndex* block_index, DBVa
-bool CoinStatsIndex::LookUpStats(const CBlockIndex* block_index, CCoinsStats& coins_stats) const
+std::optional<CCoinsStats> CoinStatsIndex::LookUpStats(const CBlockIndex* block_index) const
 {
+    CCoinsStats stats = kernel::MakeCoinStatsPrefilledWithBlockIndexInfo(block_index);
```
- [ ]

Miminlaili commented on src/index/coinstatsindex.cpp:
```diff
@@ -316,28 +317,31 @@ static bool LookUpOne(const CDBWrapper& db, const CBlockIndex* block_index, DBVa
     return db.Read(DBHashKey(block_index->GetBlockHash()), result);
 }
 
-bool CoinStatsIndex::LookUpStats(const CBlockIndex* block_index, CCoinsStats& coins_stats) const
```
- [ ]

Miminlaili reviewed (approved):
- [ ]

dongcarl commented:
Pushed 06dea7ea36eb1606cd4255e2875d0e4fe2fdd5a4 -> 660bb3066fff29bd2a16ea3adf9748d58f53363c
- Also remove `blockfilter.cpp` from `libbitcoinkernel_la_SOURCES` (Thanks past Carl for leaving a TODO!)

dongcarl commented:
I've gotten the feedback that the distinction between:

- `src/kernel/coinstats.h`, and
- `src/kernel/include/coinstats.h`

Is probably not necessary in Phase 1 of the libbitcoinkernel project and potentially confusing and lead to churn since the public one (`src/kernel/include/coinstats.h`) is likely to be drastically changed in Phase 2. I will combine them in this and future libbitcoinkernel Phase 1 PRs if there's not much opposition to that. Shouldn't affect the meat of this PR that much.

dongcarl commented:
Pushed 660bb3066fff29bd2a16ea3adf9748d58f53363c -> 8f50fe837eaa41b045c69a97617b2346d57bd879
- Addressed: https://github.com/bitcoin/bitcoin/pull/24410#issuecomment-1118984576

theuni commented on src/node/coinstats.cpp:
```diff
@@ -206,17 +202,52 @@ static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats&
+{
+    CCoinsStats stats = MakeCoinStatsPrefilledWithBlockIndexInfo(pindex);
+
+    stats.index_used = true;
```
Same comment as with `hashSerialized`. `index_used` does not seem to be a value internal to `CCoinsStats` anymore.

dongcarl replied:
Although I’ve moved some ***in***-param members out of `CCoinStats`, I think it’s outside the scope of this PR to move ***out***-param members out of `CCoinStats` but would be a worthwhile followup. This is because a proper cleanup of `CCoinStats`' out-params would likely result in a complete split of the struct itself.

Observe that:

- The hashing UTXO stats code path only ever sets the CCoinsStats members:
    - CCoinsStats::CCoinsStats
        - nHeight
        - hashBlock
    - ApplyStats
        - nTransactions
        - nTransactionOutputs
        - total_amount
        - nBogoSize
    - ComputeUTXOStatsWithHasher
        - coins_count
        - hashSerialized
        - nDiskSize
- The indexing UTXO stats code path only ever sets the CCoinsStats members:
    - CCoinsStats::CCoinsStats
        - nHeight
        - hashBlock
    - CoinstatsIndex::LookupStats
        - index_used
        - hashSerialized
        - nTransactionOutputs
        - nBogoSize
        - total_amount
        - total_subsidy
        - total_unspendable_amount
        - total_prevout_spent_amount
        - total_new_outputs_ex_coinbase_amount
        - total_coinbase_amount
        - total_unspendables_genesis_block
        - total_unspendables_bip30
        - total_unspendables_scripts
        - total_unspendables_unclaimed_rewards
- The only 6 common fields that are set:
    - nHeight
    - hashBlock
    - nTransactionOutputs
    - total_amount
    - nBogoSize
    - hashSerialized

So a proper cleanup would make it so that `CCoinStats` only contains the 6 common fields, and have other structs to contain the fields unique to each codepath.

theuni commented on src/node/coinstats.h:
```diff
@@ -76,6 +76,8 @@ std::optional<CCoinsStats> GetUTXOStats(CCoinsView* view, node::BlockManager& bl
 
 CDataStream TxOutSer(const COutPoint& outpoint, const Coin& coin);
+
+CCoinsStats MakeCoinStatsPrefilledWithBlockIndexInfo(const CBlockIndex* pindex);
```
Why not make this a ctor which takes a `CBlockIndex*` and prefills itself?

fjahr replied:
Agree, I was going to leave a nit that this is a bit verbose for my taste ;)

dongcarl replied:
Fixed as of push 8b853034a13f49b85cdf4168d9d49bd297f5b8ce

theuni commented on src/kernel/coinstats.h:
```diff
@@ -0,0 +1,88 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_KERNEL_COINSTATS_H
```
Note to other reviewers, the `node/coinstats.h` is deleted in subsequent commits, so the fact that they share a filename isn't really interesting.

theuni reviewed (commented):
Concept ACK. This is a pretty shallow review, I'm planning on following up with a more thorough one once I've explored more.

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -809,6 +808,28 @@ CoinStatsHashType ParseHashType(const std::string& hash_type_input)
+    }
+
+    auto hasher = kernel::MakeUTXOHasher(hash_type);
+    return kernel::GetUTXOStatsWithHasher(*hasher, view, blockman, interruption_point);
```
Maybe we could make an even clearer naming distinction here, I think. While we can really only "get" the stats from the index the kernel actually computes them. So the kernel function could be renamed to `ComputeUTXOStatsWithHasher`. But it's fine to leave it as is. I just found it an interesting thought and interested to hear what others think.

dongcarl replied:
Fixed as of push 8b853034a13f49b85cdf4168d9d49bd297f5b8ce

fjahr reviewed (commented):
Code review ACK 8f50fe837eaa41b045c69a97617b2346d57bd879

Very nice work. I didn't see any real issues, just some minor stylistic suggestions. Overall I think  this is a great improvement.

FWIW, I found 8c8d3f186df95b419fba8fddaeaac9b8912646dd more confusing than helpful with the functions that were introduced and removed in the following commit. But overall great work with splitting these commits to bite-sized chunks.

I will also revisit some of the tests of this code now that I looked at it for a while, but that is tangential.

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -78,6 +78,112 @@ static void ApplyHash(MuHash3072& muhash, const uint256& hash, const std::map<ui
 
+class UTXOHasher
+{
+public:
```
> Maybe give it a protected constructor?

I'm a bit confused here, is the intent just to convey in code that it's a "virtual base class" by protecting the constructor? Would a code comment suffice?

> Is there any reason to ever declare a base `UTXOHasher`?

My initial thought was that leaving the abstract `UTXOHasher` in the header (as done in commits subsequent to the one you're commenting on) could allow users to provide their specific implementation (e.g. they have some hash accelerator that we can't make use of).

This was very easily achieved by just leaving the class in the header, but I don't want to waste any time thinking about this at all so if you feel strongly about it, I'll just move it out.

dongcarl replied:
No longer relevant as of 90358d4b7402bfd4cd9b79ebf828e40f4eabd849

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -206,17 +202,52 @@ static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats&
+
+std::optional<CCoinsStats> GetUTXOStatsWithIndex(CoinStatsIndex& coin_stats_index, CCoinsView* view, BlockManager& blockman)
+{
+    CBlockIndex* pindex = WITH_LOCK(cs_main, return blockman.LookupBlockIndex(view->GetBestBlock()));
```
Fixed as of push 8b853034a13f49b85cdf4168d9d49bd297f5b8ce

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -78,6 +78,112 @@ static void ApplyHash(MuHash3072& muhash, const uint256& hash, const std::map<ui
+
+UTXOHasher::~UTXOHasher() = default;
+
+class NullHasher : public UTXOHasher {
```
Fixed as of push 8b853034a13f49b85cdf4168d9d49bd297f5b8ce

dongcarl commented on src/node/coinstats.h:
```diff
@@ -69,12 +66,10 @@ struct CCoinsStats {
 
 //! Calculate statistics about the unspent transaction output set
-bool GetUTXOStats(CCoinsView* view, node::BlockManager& blockman, CCoinsStats& stats, const std::function<void()>& interruption_point = {}, const CBlockIndex* pindex = nullptr);
+bool GetUTXOStats(CCoinsView* view, node::BlockManager& blockman, CCoinsStats& stats, CoinStatsHashType hash_type, const std::function<void()>& interruption_point = {}, const CBlockIndex* pindex = nullptr);
```
I think this is completely changed when we introduce `ComputeUTXOStatsWithHasher(UTXOHasher& hasher, ...)` with `UTXOHasher&` as the first parameter. Does that look alright?

If it still looks unclear, I'm happy to add Doxygen comments so that's it's more clear!

dongcarl replied:
I've experimented with this on top of the latest push 90358d4b7402bfd4cd9b79ebf828e40f4eabd849, but the template parameter would be based on a variable and that's not allowed in C++ (at least in clang).

michaelfolkson commented:
Spent some time trying to understand the direction of the project as a whole in the run up to today's [PR review club](https://bitcoincore.reviews/24410) and to the extent it builds on multiprocess, AssumeUTXO work. Approach ACK on the project roadmap thus far.

On this PR specifically Approach ACK, code review (limited familiarity with the coinstats code before this), no testing.

As others have said great organization of commits and commit messages (and innovative use of thumbs up/down to get reviewer reaction :) )

theStack reviewed (commented):
Concept ACK

While reviewing #24952 recently, I found that the `GetUTXOStats` interface is currently quite confusing, glad to see that many painpoints (mixing both in- and out-params in one struct, mixed hashing and indexing codepaths in one function...) are fixed with this PR. Lightly reviewed each commit, the changes are concise, well-documented in the commit descriptions and make sense to me; will look over more deeply a second time within the next days. As stated by other reviewers before, kudos for keeping the commits small and to the point 👌

dongcarl commented on src/rpc/blockchain.cpp:
```diff
@@ -922,10 +944,14 @@ static RPCHelpMan gettxoutsetinfo()
+                    prev_stats = maybe_prev_stats.value();
+                } else {
+                    throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to read UTXO set");
+                }
```
Fixed as of https://github.com/bitcoin/bitcoin/pull/24410#issuecomment-1127846221

ryanofsky reviewed (approved):
Code review ACK 0cc58453e2479a8877e2c0b463d3a7c215ceb77b.

End result looks good, but it seemed to me this PR had a lot of internal churn which was tedious to review. If function signature is changing, it's nice it if changes one time instead of one parameter at a time. If code is moving, it's nice if it moves one instead of moving multiple times. In general, unless there is a specific reason (like separating behavior changes from refactoring changes, or separating mechanical refactoring from conceptual refactoring) it's nice if each line of code changes once per PR, so the changes in the PR can be comprehended without having to think about intermediate states which turn out to be irrelevant.

But I'm venting a little. In general things look very good here.

dongcarl commented on src/node/coinstats.h:
```diff
@@ -41,8 +41,6 @@ struct CCoinsStats {
     //! The number of coins contained.
     uint64_t coins_count{0};
 
-    //! Signals if the coinstatsindex should be used (when available).
```
Fixed in 56d7d36ea7c3be0dc015bffc1d3bf8253a917abd

dongcarl commented on src/rpc/blockchain.cpp:
```diff
@@ -2320,8 +2324,11 @@ UniValue CreateUTXOSnapshot(
+        if (!maybe_stats) {
             throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to read UTXO set");
+        } else {
+            stats = maybe_stats.value();
```
Fixed in 9ec45ff9eba818ce7908344b614b550c3b276bdd

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -19,6 +19,11 @@
 
 namespace node {
+
+CCoinsStats::CCoinsStats(const CBlockIndex* pindex)
```
Fixed in 7c0064d451de00a3e53dfd8d29c3fc9e73d4d11a

dongcarl commented:
Pushed 0cc58453e2479a8877e2c0b463d3a7c215ceb77b -> abeee4b928eacb661a5829f464f0cda10da171c2
- Addressed Ryan's review comments (thanks Ryan!)

dongcarl commented:
Reviewers, I'm considering dropping the commit:

- 9dc8093189e41849e91f8425f02e8f30563dfdf3 "coinstats: Introduce and switch to UTXOHashers"

Originally I had intended for the `UTXOHashers` class to make the `GetUTXOStats` interface more extensible (one could imagine a user providing their own `UTXOHashers` implementation), but I think it might have been more confusing than not and might not be worth the trouble because YAGNI.

See discussions:
- https://github.com/bitcoin/bitcoin/pull/24410#discussion_r868360164
- https://github.com/bitcoin/bitcoin/pull/24410#discussion_r868450711
- https://github.com/bitcoin/bitcoin/pull/24410#discussion_r875408643

On the other hand, I also don't want to impose too much of a re-review burden on folks, and it really doesn't seem like it matters all that much in practice which way we go.

Let me know your opinion with a reaction to this comment:
- 👍 Drop the commit and remove `UTXOHashers`
- 😕 Keep the commit and `UTXOHashers`

Ping @mzumsande @theuni @fjahr @theStack @ryanofsky

ryanofsky reviewed (approved):
Code review ACK abeee4b928eacb661a5829f464f0cda10da171c2. Only changes since last review were implementing some suggestions from my review (avoiding CBlockIndex* constructor, consolidating stats variables, documenting index_requested param).

I'd be happy to see hasher thing simplified and `ComputeUTXOStats` not to need a hasher parameter, but also happy to see this merged as is, so I abstain from the emoji vote above!

theStack commented on src/node/coinstats.cpp:
```diff
@@ -37,45 +37,110 @@ CDataStream TxOutSer(const COutPoint& outpoint, const Coin& coin) {
+    return std::make_unique<NullHasher>();
+}
+
+class SHA256DHasher final : public UTXOHasher
```
naming nit: I found the name `SHA256DHasher` a bit too specific and misleading, at first glance one could think that this is a helper for generally computing a SHA256D hash. IMHO the important criteria for naming is not what concrete hash function a UTXO hasher interally uses (e.g., MuHash also uses SHA256 for finalizing, and ChaCha20 for kind of "hashing" UTXOs to big integers via a stream cipher), but rather how it works on a more abstract conceptual level and in what way it commits to UTXOs etc. Not saying that's is easy to come up with a better name, but I'd prefer to keep `UTXO` in it, prefixed with a name resembling the output fields of the `gettxoutsetinfo` fields, e.g. `Legacy(Serialized)UTXOHasher` or `MuHashUTXOHasher`.

(Probably this comment will be obsolete anyway though if commit 9dc8093189e41849e91f8425f02e8f30563dfdf3 gets dropped)

dongcarl replied:
Yeah that's a good point, in any case I think it's no longer relevant as of 90358d4b7402bfd4cd9b79ebf828e40f4eabd849

theStack reviewed (commented):
>Reviewers, I'm considering dropping the commit:
>
>https://github.com/bitcoin/bitcoin/commit/9dc8093189e41849e91f8425f02e8f30563dfdf3 "coinstats: Introduce and switch to UTXOHashers"
>
>Originally I had intended for the UTXOHashers class to make the GetUTXOStats interface more extensible (one could imagine a user providing their own UTXOHashers implementation), but I think it might have been more confusing than not and might not be worth the trouble because YAGNI.

In my first round of (very light) code-review I found the introduction of the `UTXOHasher` abstract interface quite a good idea, but then on the other hand ryanofsky brings up really strong counter-arguments (https://github.com/bitcoin/bitcoin/pull/24410#discussion_r875408643) that are hard to refute. Particularly "more heap allocations" and "more complicated kernel API" are valid points that I didn't consider (the pinning of the hash size to 256 bits seems less of a problem to me though), so that convinced me to vote for dropping the commit.

dongcarl commented:
Pushed abeee4b928eacb661a5829f464f0cda10da171c2 -> 90358d4b7402bfd4cd9b79ebf828e40f4eabd849
- Implemented https://github.com/bitcoin/bitcoin/pull/24410#issuecomment-1132161124
- A few minor cleanups (scripted-diff simplification, extraneous comments)

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -37,45 +37,110 @@ CDataStream TxOutSer(const COutPoint& outpoint, const Coin& coin) {
+public:
+    virtual void Prepare(const uint256& hash_block){};
+    virtual void Apply(const uint256& hash, const std::map<uint32_t, Coin>& outputs){};
+    virtual uint256 Finalize() = 0;
```
Resolved as of 90358d4b7402bfd4cd9b79ebf828e40f4eabd849

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -78,6 +78,112 @@ static void ApplyHash(MuHash3072& muhash, const uint256& hash, const std::map<ui
+{
+    switch (hash_type) {
+    case(CoinStatsHashType::HASH_SERIALIZED): {
+        return MakeSHA256DHasher();
```
No longer relevant as of 90358d4b7402bfd4cd9b79ebf828e40f4eabd849

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -241,31 +240,19 @@ static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats&
     }
 
-    FinalizeHash(hash_obj, stats);
+    stats.hashSerialized = hasher.Finalize();
```
No longer relevant as of 90358d4b7402bfd4cd9b79ebf828e40f4eabd849

dongcarl commented on src/node/coinstats.h:
```diff
@@ -69,12 +66,10 @@ struct CCoinsStats {
     //! Total cumulative amount of coins lost due to unclaimed miner rewards up to and including this block
     CAmount total_unspendables_unclaimed_rewards{0};
-
-    CCoinsStats(CoinStatsHashType hash_type) : m_hash_type(hash_type) {}
```
Yup! Feel free to reopen this if it's still a problem.

mzumsande commented on src/node/coinstats.cpp:
```diff
@@ -144,22 +145,32 @@ static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats&
     } // no default case, so the compiler can warn about missing cases
-    assert(false);
+
+    if (!success) {
```
2e1ff3915cab0e902195514f04378f74e3453610 (coinstats: Return purely out-param CCoinsStats)

This gives me a `-Wmaybe-uninitialized` compiler warning for `success` (although this is obviously impossible).

dongcarl replied:
Hmmm, do you think we need to do anything here? I'm not sure how best to fix this since we want to keep it so that the compiler can warn about missing cases

mzumsande replied:
Maybe just initialise success with `true` or `false` to appease the compiler?
I guess the catch-22 ist that if we had multiple warnings like this on master, we'd easily miss legitimate warnings such as the one about missing cases, which could then lead to `success` actually being uninitialized in this spot...
But I don't feel strongly about this.

MarcoFalke replied:
It is not UB (I think) to pass any underlying integral value as an `enum class`, so uninitialized `success` can in theory happen in a "well-formed" program.

A few alternative fixes:

* Initialize success with `false`
* Keep the `assert(false)` and inline all return statements into the case statments: `if (!Calc(...)) return std::nullopt; return stats;`
* Keep the `assert(false)`, but wrap it with the whole `switch (hash_type)` scope into an inlined lambda
* Something else ... ?

ryanofsky replied:
In commit "coinstats: Return purely out-param CCoinsStats" (f735c92edc08047bdbe0f47ea0e7c319bcc62eeb)

Agree with Marco, I think. I don't understand why you wouldn't initialize success to false. If function is called with `(CoinStatsHashType)123` compiler is correct to warn that it has undefined behavior. It seems like it would be best to assert false in this case, second best to return nullopt if asserting is inconvenient, and third best to have undefined behavior.

dongcarl replied:
Fixed in 524463daf6a10b20a4e20116a68101a684929eda using a lambda, thanks for the help here folks!

mzumsande reviewed (commented):
Review the code one more time - this looks great! 
Also did some light testing that `gettxoutsetinfo` still works as before.

Two points below that are minor but would be worth fixing imo.

ryanofsky commented on src/node/coinstats.cpp:
```diff
@@ -193,4 +187,37 @@ static void FinalizeHash(MuHash3072& muhash, CCoinsStats& stats)
+        }
+    }
+
+    return ComputeUTXOStats(hash_type, view, blockman, interruption_point);
```
In commit "coinstats: Separate hasher/index lookup codepaths" (e4b2969b54a3ddba507de1aeb17b6765ca1f102b)

Behavior precedes this PR, but it seems like `pindex` just gets ignored in this case. It might be nice to `assert(pindex == nullptr)` here or else have a comment "// pindex is ignored if coinstats index is not available or wasn't requested. Coinstats for the view are returned instead. This is ok because \[reasons]"

dongcarl replied:
Fixed in d03f4443e8deb477d0fd4b27056eaeec8a615ca3, thanks!

ryanofsky commented on src/node/coinstats.cpp:
```diff
@@ -188,33 +188,15 @@ static void FinalizeHash(MuHash3072& muhash, CCoinsStats& stats)
-    return stats;
-}
-
-std::optional<CCoinsStats> LookupUTXOStatsWithIndex(CoinStatsIndex& coin_stats_index, CCoinsView* view, BlockManager& blockman)
```
In commit "Move logic from LookupUTXOStatsWithIndex to CoinStatsIndex::LookUpStats" (528472c420dccbeb673dd77b64d327ab5271f23b)

This might be an example of unnecessary churn. Not a big deal, but why go through trouble of adding these LookupUTXOStatsWithIndex functions in e4b2969b54a3ddba507de1aeb17b6765ca1f102b when they will be deleted in 528472c420dccbeb673dd77b64d327ab5271f23b

ryanofsky reviewed (approved):
Code review ACK 90358d4b7402bfd4cd9b79ebf828e40f4eabd849. I rereviewed from scratch but main change since last review seems to be not adding hasher objects. All looks good, feel free to ignore minor comments.

theStack commented on src/rpc/blockchain.cpp:
```diff
@@ -808,6 +808,30 @@ CoinStatsHashType ParseHashType(const std::string& hash_type_input)
+ *
+ * @param[in] index_requested Signals if the coinstatsindex should be used (when available).
+ */
+std::optional<kernel::CCoinsStats> GetUTXOStats(CCoinsView* view, node::BlockManager& blockman,
```
nit: could be made static, since this function is not used outside this module
```suggestion
static std::optional<kernel::CCoinsStats> GetUTXOStats(CCoinsView* view, node::BlockManager& blockman,
```

dongcarl replied:
Fixed in e6e2b165fd377806e257fbb5378e22067a9a3957, thanks!

theStack reviewed (approved):
Code-review ACK 90358d4b7402bfd4cd9b79ebf828e40f4eabd849 🏞️ 

(I'm aware that this needs rebase, but re-ACKing should be fairly trivial)

dongcarl commented:
Pushed 90358d4b7402bfd4cd9b79ebf828e40f4eabd849 -> e6e2b165fd377806e257fbb5378e22067a9a3957

- Rebased over master
- Addressed:
  - https://github.com/bitcoin/bitcoin/pull/24410#discussion_r878378791
  - https://github.com/bitcoin/bitcoin/pull/24410#discussion_r878475165
  - https://github.com/bitcoin/bitcoin/pull/24410#discussion_r878520316

dongcarl commented on src/node/coinstats.cpp:
```diff
@@ -94,24 +99,11 @@ static void ApplyStats(CCoinsStats& stats, const uint256& hash, const std::map<u
 //! Calculate statistics about the unspent transaction output set
 template <typename T>
-static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats& stats, T hash_obj, const std::function<void()>& interruption_point, const CBlockIndex* pindex, CoinStatsHashType& hash_type, bool index_requested)
+static bool ComputeUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats& stats, T hash_obj, const std::function<void()>& interruption_point)
```
Fixed in d03f4443e8deb477d0fd4b27056eaeec8a615ca3, thanks!

dongcarl commented:
Pushed e6e2b165fd377806e257fbb5378e22067a9a3957 -> e42583f928d5461827a946db3baa7b3da9487253
- Removed over-eager assertion: `GetUTXOStats` is sometimes called in `rpc/blockchain.cpp` with a non-null `pindex` even when we aren't going to use the index. In all such cases the `pindex` represents `view`'s best block.

theStack reviewed (approved):
re-ACK e42583f928d5461827a946db3baa7b3da9487253

ryanofsky commented on src/node/coinstats.cpp:
```diff
@@ -193,4 +187,44 @@ static void FinalizeHash(MuHash3072& muhash, CCoinsStats& stats)
+    //
+    // This is ok because we never call GetUTXOStats in a way where we can't use
+    // the coinstatsindex and the specified pindex is not also the view's best
+    // block.
```
In commit "coinstats: Separate hasher/index lookup codepaths" (99b7dced926eeba1ac0b181feb18ebb7ce35504e)

Seems like it would be easy for code calling this function to be unaware of this assumption. Could check it with
`assert(!pindex || pindex->GetBlockHash() == view->GetBestBlock());`

dongcarl replied:
Right! Fixed in 1352e410a5b84070279ff28399083cb3ab278593

ryanofsky reviewed (approved):
Code review ACK e42583f928d5461827a946db3baa7b3da9487253. Looks good, I left new suggestions but they are not important. Changes since last review were all minor: adding comment, adding static, dropping unused parameter

dongcarl commented:
Push e42583f928d5461827a946db3baa7b3da9487253 -> c7e8803e47eadc68e46e9384a7e2c4543404cd1b
- Addressed
  - https://github.com/bitcoin/bitcoin/pull/24410#discussion_r879649251
  - https://github.com/bitcoin/bitcoin/pull/24410#discussion_r879660117

dongcarl commented:
Pushed c7e8803e47eadc68e46e9384a7e2c4543404cd1b -> 664a14ba7ccb40aa82d35a59831acd35db1897a6
- Use `CHECK_NONFATAL` instead of `assert` in `GetUTXOStats` after moving to `rpc/blockchain.cpp`

laanwj commented:
Code review ACK 664a14ba7ccb40aa82d35a59831acd35db1897a6

laanwj commented on src/kernel/coinstats.cpp:
```diff
@@ -141,25 +131,36 @@ static bool GetUTXOStats(CCoinsView* view, BlockManager& blockman, CCoinsStats&
+
+    bool success = [&]() -> bool {
+        switch (hash_type) {
+        case(CoinStatsHashType::HASH_SERIALIZED): {
```
nit: I've never seen this syntax before, why `case(CoinStatsHashType::HASH_SERIALIZED)` instead of `case CoinStatsHashType::HASH_SERIALIZED`?

laanwj replied:
Oh, that was already in the original code, never mind.

theStack reviewed (commented):
post-merge re-ACK 664a14ba7ccb40aa82d35a59831acd35db1897a6
